httpbakery/visitor.go

```diff
@@ -0,0 +1,108 @@
+package httpbakery
+
+import (
+	"net/http"
+	"net/url"
+
+	"github.com/juju/httprequest"
+	"gopkg.in/errgo.v1"
+)
+
+var ErrMethodNotSupported = errgo.New("interaction method not supported")
+
+// Visitor represents a handler that can handle ErrInteractionRequired errors
+// from a client's discharge request. The methodURLs parameter to
+// VisitWebPage holds a set of possible ways to complete the
+// discharge request. When called directly from Client, this
+// will contain only a single entry with the UserInteractionMethod key,
+// specifying that the associated URL should be opened
+// in a web browser for the user to interact with.
+//
+// See FallbackVisitor for a way to gain access to
+// alternative methods.
+//
+// A Visitor implementation may return ErrMethodNotSupported
+// if it cannot handle any of the supplied methods.
+type Visitor interface {
+	VisitWebPage(client *Client, methodURLs map[string]*url.URL) error
+}
+
+const (
+	// UserInteractionMethod is the methodURLs key
+	// used for a URL that should be visited in a user's
+	// web browser. This is also the URL that can
+	// be used to fetch the available login methods (with
+	// an appropriate Accept header).
+	UserInteractionMethod = "Interactive"
+)
+
+type multiVisitor struct {
+	supportedMethods []Visitor
+}
+
+// NewMultiVisitor returns a Visitor that queries the
+// discharger for available methods and then tries each
+// of the given visitors in turn until one succeeds
+// or fails with an error cause other than ErrMethodNotSupported.
+func NewMultiVisitor(methods ...Visitor) Visitor {
+	return &multiVisitor{
+		supportedMethods: methods,
+	}
+}
+
+func (v multiVisitor) VisitWebPage(client *Client, methodURLs map[string]*url.URL) error {
+	u := methodURLs[UserInteractionMethod]
+	if u == nil {
+		return errgo.Newf("cannot get interaction methods because no %q URL found", UserInteractionMethod)
+	}
+	if urls, err := GetInteractionMethods(client, u); err == nil {
+		// We succeeded in getting the set of interaction methods from
+		// the discharger. Use them.
+		methodURLs = urls
```

Comment:
I would like to check here that methodURLs contains UserInteractionMathods and merge it in if it doesn't. This mkaes the whole thing backwards compatible with the currently deployed IdM.